app/src/main/java/com/stevesoltys/seedvault/restore/install/ApkSplitCompatibilityChecker.kt

```diff
@@ -0,0 +1,134 @@
+
+    private val abiMap = mapOf(
+        "armeabi" to "armeabi",
+        "armeabi_v7a" to "armeabi-v7a",
+        "arm64_v8a" to "arm64-v8a",
+        "x86" to "x86",
+        "x86_64" to "x86_64",
+        "mips" to "mips",
+        "mips64" to "mips64"
+    )
+    private val densityMap = mapOf(
+        "ldpi" to LDPI_VALUE,
+        "mdpi" to MDPI_VALUE,
+        "tvdpi" to TVDPI_VALUE,
+        "hdpi" to HDPI_VALUE,
+        "xhdpi" to XHDPI_VALUE,
+        "xxhdpi" to XXHDPI_VALUE,
+        "xxxhdpi" to XXXHDPI_VALUE
+    )
+
+    /**
+     * Returns true if the list of splits can be considered compatible with the current device,
+     * and false otherwise.
+     */
+    fun isCompatible(deviceName: String, splitNames: Collection<String>): Boolean {
+        val unknownAllowed = deviceInfo.areUnknownSplitsAllowed(deviceName)
+        return splitNames.all { splitName ->
+            // all individual splits need to be compatible (which can be hard to judge by name only)
+            isCompatible(deviceName, unknownAllowed, splitName)
+        }
+    }
+
+    private fun isCompatible(
+        deviceName: String,
+        unknownAllowed: Boolean,
+        splitName: String
+    ): Boolean {
+        val index = splitName.indexOf(CONFIG_PREFIX)
+        // If this is not a standardized config split
+        if (index == -1 && unknownAllowed) {
+            // we assume that it will work, as it is most likely a dynamic feature module
+            Log.v(TAG, "Not a config split '$splitName'. Assuming it is ok.")
+            return true
+        } else if (index != 0 && !unknownAllowed) { // not a normal config split at all
+            // we refuse it, since unknown splits are not allowed
+            Log.v(TAG, "Not a config split '$splitName'. Not allowed.")
+            return false
+        }
+
+        val name = splitName.substring(index + CONFIG_LENGTH)
+
+        // Check if this is a known ABI config
+        if (abiMap.containsKey(name)) {
+            // The ABI split must be supported by the current device
+            return isAbiCompatible(name)
+        }
+
+        // Check if this is a known screen density config
+        densityMap[name]?.let { splitDensity ->
+            // The split's density must not be much lower than the device's.
+            return isDensityCompatible(splitDensity)
+        }
+
+        // Check if this is a language split
+        if (deviceInfo.isSupportedLanguage(name)) {
+            // accept all language splits as an extra language should not break things
+            return true
+        }
+
+        // At this point we don't know what to make of that split
+        return if (deviceInfo.isSameDevice(deviceName)) {
+            // so we only allow it if it came from the same device
+            Log.v(TAG, "Unhandled config split '$splitName'. Came from same device.")
+            true
+        } else {
+            Log.w(TAG, "Unhandled config split '$splitName'. Not allowed.")
```

Comment:
Would we want to allow these when `unknownAllowed` is true?

Reply:
I think this is too risky as there can be lots of crazy splits and we have no way to check compatibility here.

Reply:
Ok so unknown splits are ok if enabled, unless they're unknown config.* splits, since for those we have a well-defined list of what we accept. Right?

Reply:
Right.